aten/src/ATen/native/ConvolutionMM2d.cpp

```diff
@@ -410,34 +417,45 @@ std::tuple<Tensor&, Tensor&> slow_conv2d_forward_out_cpu(
                   n_input_plane * kernel_height * kernel_width,
                   output_height * output_width});
   }
-  output.resize_({batch_size, n_output_plane, output_height, output_width});
 
-  at::parallel_for(0, batch_size, 0, [&](int64_t start, int64_t end) {
-    NoGradGuard no_grad;
-    AutoDispatchBelowADInplaceOrView non_variable_type_mode;
-    for (int64_t t = start; t < end; t++) {
-      Tensor input_t = input[t].unsqueeze(0);
-      Tensor output_t = output[t];
-      Tensor finput_t = finput[t];
-      slow_conv2d_update_output_frame(
-          input_t,
-          output_t,
-          weight_2d,
-          bias,
-          finput_t,
-          kernel_height,
-          kernel_width,
-          stride_height,
-          stride_width,
-          pad_height,
-          pad_width,
-          n_input_plane,
-          input_height,
-          input_width,
-          n_output_plane,
-          output_height,
-          output_width);
-    }
+  output.resize_({batch_size, n_output_plane, output_height, output_width});
+  if (bias.defined()) {
+    output.copy_(bias.reshape({-1, 1, 1}));
+  }
+  TORCH_CHECK(output.is_contiguous() && finput.is_contiguous(),
+              "slow_conv2d output tensors must be contiguous");
+
+  AT_DISPATCH_ALL_TYPES_AND(kBFloat16, input.scalar_type(), "slow_conv2d_cpu", [&]{
```

Comment:
I don't think ALL_TYPES and bfloat16 makes sense here. In contrast to addmm, fast bfloat16 support is not enabled on direct calls to at::cpublas::gemm, and in CPUBlas.cpp there are specializations only for float, double, complex float and double and int64, the rest go to extremely slow `gemm_stub`.

Reply:
These are the dtypes that conv2d supports. As limited by Unfold2d, which is only used in this code:
https://github.com/pytorch/pytorch/blob/1d681c1ab2e335de2ec32995dc2aa4e57dd873e7/aten/src/ATen/native/cpu/Unfold2d.cpp#L287-L288

Although this did make me realize that supporting `ALL_TYPES` for backward functions doesn't make any sense. So I've changed those PRs to use `FLOATING_TYPES_AND(kBFloat16`

>In contrast to addmm, fast bfloat16 support is not enabled on direct calls to at::cpublas::gemm

I assume this is #64619. That seems like a weird choice to me. Do you know why they special case addmm instead of gemm?

Reply:
Probably because it allowed them to have a single `matmul` function instead of separate `gemm` and batched gemm, also iirc their `matmul` can handle broadcasting in the addition part, that gemm cannot, and integrating at `gemm` level would not have allowed them to take advantage of that (if `b` in `ax+b` has to be broadcasted, we first explicitly expand it and them call gemm)